src/services/redeem_and_unstake/ConfirmRedeemIntentService.ts

```diff
@@ -0,0 +1,193 @@
+
+  private originWeb3: Web3;
+
+  private auxiliaryWeb3: Web3;
+
+  private gatewayAddress: string;
+
+  private coGatewayAddress: string;
+
+  private originWorkerAddress: string;
+
+  /**
+   * Constructor of class ConfirmRedeemIntentService;
+   *
+   * @param messageRepository Instance of message repository.
+   * @param messageTransferRequestRepository Instance of request repository.
+   * @param originWeb3 Instance of origin chain web3.
+   * @param auxiliaryWeb3 Instance of auxiliary chain web3.
+   * @param gatewayAddress Origin chain gateway address.
+   * @param coGatewayAddress Auxiliary chain gateway address.
+   * @param originWorkerAddress Auxiliary chain worker address.
+   */
+  public constructor(
+    messageRepository: MessageRepository,
+    messageTransferRequestRepository: MessageTransferRequestRepository,
+    originWeb3: Web3,
+    auxiliaryWeb3: Web3,
+    gatewayAddress: string,
+    coGatewayAddress: string,
+    originWorkerAddress: string,
+  ) {
+    super();
+
+    this.messageRepository = messageRepository;
+    this.messageTransferRequestRepository = messageTransferRequestRepository;
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.gatewayAddress = gatewayAddress;
+    this.coGatewayAddress = coGatewayAddress;
+    this.originWorkerAddress = originWorkerAddress;
+  }
+
+  /**
+   * This method reacts on changes when GatewayProven entity is received.
+   *
+   * Gateway model first object is selected because of update argument interface. From
+   * ProveGatewayHandler always single Gateway model is passed.
+   *
+   * Messages to be send for confirmation is fetched and confirmRedeemIntent is called.
+   *
+   * @param gateway List of Gateway models.
+   */
+  public async update(gateway: Gateway[]): Promise<void> {
+    Logger.debug('Confirm redeem intent service invoked');
+    assert(
+      gateway.length === 1,
+      'There should be only 1 gateway record',
+    );
+    const provenGateway: Gateway = gateway[0];
+    const messages: Message[] = await this.messageRepository.getMessagesForConfirmation(
+      provenGateway.gatewayAddress,
+      provenGateway.lastRemoteGatewayProvenBlockHeight as BigNumber,
+      MessageDirection.AuxiliaryToOrigin,
+    );
+
+    await this.confirmRedeemIntent(provenGateway, messages);
+  }
+
+  /**
+   * Collects all ConfirmRedeemIntent promises and transaction is sent.
+   *
+   * @param gateway Instance of Gateway model object.
+   * @param messages List of message models.
+   */
+  private async confirmRedeemIntent(
+    gateway: Gateway,
+    messages: Message[],
+  ): Promise<Record<string, string>> {
+    if (messages.length === 0) {
+      return {};
+    }
+
+    const proofGenerator = new ProofGenerator(
+      this.auxiliaryWeb3,
+    );
+
+    const transactionHashes: Record<string, string> = {};
+
+    const promises = messages
+      .filter((message): boolean => message.isValidSecret())
```

Comment:
In map we will not be able to do because output/return of a map iteration should always be a valid promise. Also this pattern we have followed at multiple flows which is quite readable.

Reply:
we can always return Promise.resolve() 
but micros optimisations I guess. can live without it